Place Event (Implements #5054)

coaster3000 opened the pull request:
Pull request changes the BlockPlace event to allow more use cases in modding. It adds the ability to change the default behavior of block placement for actions such Falling block being placed, Endermen Block Placement, and any other areas which call the world's placement check method.

Changes have been made in various locations that call the world's placement check method due to missing data when the method was called. All calls that have been edited were edited to add data to the final parameter of the call. An entity parameter that states what entity placed the block always had null passed to it for an unknown reason. It has been updated to pass known entity data.

The locations of these changes includes:
- EntityFallingBlock
- ItemBlock
- ItemBlockSpecial
- ItemRedstone

Additionally, changes have been made to the ForgeEventFactory to keep method naming as well as documentation consistent with the new BlockPlace event behavior. The old methods have been redirected to the new methods and deprecated. All references within the existing code have been updated to use the new method. A deprecation note has been added, stating to use the new method.

--------------

The PlaceEvent test mod code has been updated to show EntityFallingBlock placements. The test mod continues to remain disabled by default due to performance reasons. The performance hit of the test does increase with these changes due to the effect applied upon block placement for falling blocks.

--------------
EDIT:
> 
> Unfortunately the changes introduces the possibility for null pointer exceptions to occur within mods if they use the event's `getPlayer()` method without any checks. For this reason, it has been deprecated and marked with `@Nullable`.
> 
> In addition to the deprecation of `getPlayer()`, the `getHand()` method has been updated with the `@Nullable` annotation due to the possibility returning null.

This is no longer the case.

coaster3000 committed:
Replace null with proper entity data

JamiesWhiteShirt commented:
This breaks binary compatibility with the changes to the event constructors, and the changes to nullability and firing contexts will definitely break mods that subscribe to the event. Changing and deprecating the methods is not the way to go.

You could perhaps shim a class between `PlaceEvent` and `BlockEvent` that handles any kind of entity, or make a more generic place event class on the side. If it gets messy we can probably have it cleaned up in the next breaking build.

I would like to add that I have a use case for these changes. I need to be able to prevent placement of blocks that would collide with a physical structure that is not represented as a block or an entity. This becomes more achievable with the additional cases covered here.

While you are already working on this, I would like to request a way to handle place events caused by players on the client. If the event is cancelled on the server, it causes the block to flicker while the client is waiting for the server to inform it that the placement was canceled. If a mod could reasonably predict the outcome on the client, it could cancel the event on the client to prevent the flicker.

[Demonstration of use case and flicker. It gets worse with ping.](https://gfycat.com/UnkemptUniqueFinnishspitz)

coaster3000 commented:
> This breaks binary compatibility with the changes to the event constructors, and the changes to nullability and firing contexts will definitely break mods that subscribe to the event. Changing and deprecating the methods is not the way to go.

I had a feeling that changes in nullability might not be acceptable and was exploring some options for this. I was looking to create a separate event for this, but had difficulty naming it at the time. 

> While you are already working on this, I would like to request a way to handle place events caused by players on the client. If the event is cancelled on the server, it causes the block to flicker while the client is waiting for the server to inform it that the placement was canceled. If a mod could reasonably predict the outcome on the client, it could cancel the event on the client to prevent the flicker.

I will definitely try and look at that. It might be more suitable in a separate pull request depending on what is changed.  

Do we have any issues made for the flickering issue? I would imagine this issue is present before this pull request. Just that the issue is not as frequent. 

EDIT: I am creating the shim classes now. Would it be appropriate to leave a todo note to update the class names on the breaking update?

JamiesWhiteShirt commented:
> Do we have any issues made for the flickering issue? I would imagine this issue is present before this pull request. Just that the issue is not as frequent.

PlaceEvent was introduced in #1367, and there was no mention of the event only being fired on the server. [It doesn't appear to be mentioned anywhere in the issue tracker either.](https://github.com/MinecraftForge/MinecraftForge/issues?utf8=%E2%9C%93&q=PlaceEvent)

It would seem the issue is perhaps not frequent, or people don't care about the flicker, or I'm just going about it the wrong way.

coaster3000 commented:
I have done some searching on my end and have found no tickets made for the issue. I believe I may have run across a section of code that checked isRemote flags and skipped the event if it was running on client side.

The patch involved with itemPlacement use is https://github.com/MinecraftForge/MinecraftForge/blob/2786cd279cd8feb5965060da27f141d7c8ccf1e5/patches/minecraft/net/minecraft/item/ItemStack.java.patch#L80 which is ignoring client side execution. The method being called is not safe to run on the client.

The code for that method is here:
https://github.com/MinecraftForge/MinecraftForge/blob/2786cd279cd8feb5965060da27f141d7c8ccf1e5/src/main/java/net/minecraftforge/common/ForgeHooks.java#L860-L949

Upon some quick searching with my ide, I found that only one method calls that function and it checks for the current side before running the hook. I believe that this whole issue should have a new issue assigned and a separate pull request made because it requires changing the behavior of the existing events. I doubt mod developers will expect the event to suddenly execute on client side.

coaster3000 commented:
While debugging, I came across some things that need to be fixed before this can be considered ready for merging. One of my existing changes appears to be firing the Event twice on player block placement.

coaster3000 commented on patches/minecraft/net/minecraft/world/World.java.patch:
```diff
@@ -857,12 +857,17 @@
+         else if (iblockstate1.func_185904_a() == Material.field_151594_q && p_190527_1_ == Blocks.field_150467_bQ)
+         {
+-            return true;
++            return !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(p_190527_5_, net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(this, p_190527_2_), p_190527_4_, null).isCanceled();
```
This change causes PlaceEvent to execute twice. Certain onItemUse calls already fire the event.

coaster3000 committed:
Create shim classes between BlockEvent, PlaceEvent, and MultiPlaceEvent

coaster3000 commented:
Have two more things that need to be ironed out before merging.
I need to write a new test mod that distinguishes events.

I also am currently investigating a MultiBlockPlace event that is being fired on chest placements.

bs2609 commented:
https://github.com/MinecraftForge/MinecraftForge/pull/3730#issuecomment-282515653

coaster3000 committed:
Add test mod to handle the BasePlaceEventTest

This mod handles BlockPlace events that are fired by falling block entities. When enabled, the event will cause all falling block entities to not be placed on the ground.

coaster3000 commented:
I would like to tackle the multiplace firing issue in another pull request. Just going to have this reviewed for merge at this point.

coaster3000 commented:
I had recently changed out my GPG keys. Commits will show up as unverified since the key has been removed from my account. Any new commits will be using my new key.

coaster3000 commented:
Still waiting review, but it seems I now have new conflicts that must be resolved first.

tterrag1098 commented:
> Unfortunately the changes introduces the possibility for null pointer exceptions to occur within mods if they use the event's getPlayer() method without any checks. For this reason, it has been deprecated and marked with @Nullable.
>In addition to the deprecation of getPlayer(), the getHand() method has been updated with the @Nullable annotation due to the possibility returning null.

I don't see this as the case anymore, has it changed since you wrote that?

tterrag1098 commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -166,25 +168,53 @@ public void setExpToDrop(int exp)
      */
     @Cancelable
-    public static class PlaceEvent extends BlockEvent
+    public static class BasePlaceEvent extends BlockEvent
```
"Base" is not very specific here, maybe use "Entity" ?
```suggestion
    public static class EntityPlaceEvent extends BlockEvent
```

coaster3000 replied:
Change has been made.

tterrag1098 commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -166,25 +168,53 @@ public void setExpToDrop(int exp)
+     * If a Block Place event is cancelled, the block will not be placed.
+     */
+    @Cancelable
+    public static class PlaceEvent extends BasePlaceEvent
```
Deprecate this and mark for removal in 1.13, no need to have two events for almost the same purpose other than backwards compat. In 1.13 this can just be an `@Nullable getPlayer()` or `Optional<EntityPlayer>` return or something along those lines.
```suggestion
    @Deprecated // Remove in 1.13
    public static class PlaceEvent extends BasePlaceEvent
```
Same applies to the multi version.

tterrag1098 commented on patches/minecraft/net/minecraft/item/ItemRedstone.java.patch:
```diff
@@ -0,0 +1,11 @@
+         ItemStack itemstack = p_180614_1_.func_184586_b(p_180614_4_);
+ 
+-        if (p_180614_1_.func_175151_a(blockpos, p_180614_5_, itemstack) && p_180614_2_.func_190527_a(p_180614_2_.func_180495_p(blockpos).func_177230_c(), blockpos, false, p_180614_5_, (Entity)null) && Blocks.field_150488_af.func_176196_c(p_180614_2_, blockpos))
++        if (p_180614_1_.func_175151_a(blockpos, p_180614_5_, itemstack) && p_180614_2_.func_190527_a(p_180614_2_.func_180495_p(blockpos).func_177230_c(), blockpos, false, p_180614_5_, p_180614_1_) && Blocks.field_150488_af.func_176196_c(p_180614_2_, blockpos))
```
Are you 100% sure that changes like this do not impact vanilla behavior?

coaster3000 replied:
> 
> 
> Are you 100% sure that changes like this do not impact vanilla behavior?

I have not seen any changes in vanilla behavior in the test I have done; However, I cannot say I have done every possible test for it myself. 

When I initially examined that segment of code when vanilla, it always pushed null values in that argument that were never used. I was very confused as to why it did that in the first place because of its lack of use. So to prevent any breaking changes, I decided to utilize the parameter to help fill in data for the event.

PS: Because it has been so long since this review, I may not remember the entire reasoning behind some of the changes. I am very busy and it will take some time to redo the edits to not have any merge conflicts as well.

coaster3000 commented:
> > Unfortunately the changes introduces the possibility for null pointer exceptions to occur within mods if they use the event's getPlayer() method without any checks. For this reason, it has been deprecated and marked with @nullable.
> > In addition to the deprecation of getPlayer(), the getHand() method has been updated with the @nullable annotation due to the possibility returning null.
> 
> I don't see this as the case anymore, has it changed since you wrote that?

No, that was fixed with the shim class implementation when JamiesWhiteShirt had commented that the original changes broke binary compatibility. However, it would be preferable to change it from players to optional entities since players are not the only entities that can place blocks.

coaster3000 committed:
Implement event hooks for block placement.

coaster3000 committed:
Rename Events to match their purpose better.

BaseMultiPlaceEvent changed to EntityMultiPlaceEvent.
BasePlaceEvent changed to EntityPlaceEvent.

coaster3000 committed:
Deprecate PlaceEvent for removal in 1.13

coaster3000 commented:
Not stale, waiting for response.

coaster3000 commented:
The branch is not stale, awaiting response.

(Deleted and reposted comment but as I deleted it, bot just cleared the mark... I thought the bot missed the webhook or something.)

tterrag1098 commented:
What was the purpose of this change? https://github.com/MinecraftForge/MinecraftForge/commit/f3d364397dec1cea9ab2d732b158ea4122e9aa5c#diff-0facadd57a1c485eed926033d315a5f0L907

coaster3000 commented:
> 
> 
> What was the purpose of this change? [f3d3643#diff-0facadd57a1c485eed926033d315a5f0L907](https://github.com/MinecraftForge/MinecraftForge/commit/f3d364397dec1cea9ab2d732b158ea4122e9aa5c#diff-0facadd57a1c485eed926033d315a5f0L907)

Apologies @tterrag1098, if I am reading the patch file correctly, that is removing one of the checks the game does. That was not an intentional change, the only change in the world file should have been the addition of the place event being fired and tested. I am unsure how that change was introduced while implementing the event within that file.

tterrag1098 commented:
It's already been changed back, see #5651
